bridges/MastodonBridge.php

```diff
@@ -17,74 +23,142 @@ class MastodonBridge extends FeedExpander {
-		}
-
-		// Check if it's a initial toot or a response
-		if($this->getInput('norep') && preg_match('/^@.+/', trim($content->plaintext))) {
-			return null;
+	public function getName() {
+		switch($this->queriedContext) {
+			case 'By username':
+				return $this->getInput('canusername');
+			default: return parent::getName();
 		}
-
-		return $item;
 	}
 
-	private function getInstance(){
+	private function getInstance() {
 		preg_match('/^@[a-zA-Z0-9_]+@(.+)/', $this->getInput('canusername'), $matches);
 		return $matches[1];
 	}
 
-	private function getUsername(){
+	private function getUsername() {
 		preg_match('/^@([a-zA-Z_0-9_]+)@.+/', $this->getInput('canusername'), $matches);
 		return $matches[1];
 	}
 
 	public function getURI(){
-		if($this->getInput('canusername'))
-			return 'https://' . $this->getInstance() . '/@' . $this->getUsername() . '.rss';
+		if($this->getInput('canusername')) {
+			// We parse webfinger to make sure the URL is correct. This is mostly because
+			// MissKey uses user ID instead of the username in the endpoint, and also to
+			// be compatible with future ActivityPub implementations.
+			$resource = 'acct:' . $this->getUsername() . '@' . $this->getInstance();
+			$webfingerUrl = 'https://' . $this->getInstance() . '/.well-known/webfinger?resource=' . $resource;
+			$webfingerHeader = array(
+				'Content-Type: application/jrd+json'
+			);
+			$webfinger = json_decode(getContents($webfingerUrl, $webfingerHeader), true);
+			foreach ($webfinger['links'] as $link) {
+				if ($link['type'] === 'application/activity+json') {
+					return $link['href'];
+				}
+			}
+		}
 
 		return parent::getURI();
 	}
 
-	public function collectData(){
-		return $this->collectExpandableDatas($this->getURI());
+	public function collectData() {
+		$url = $this->getURI() . '/outbox?page=true';
+		$content = json_decode(getContents($url, self::AP_HEADER), true);
+		if ($content['id'] === $url) {
+			foreach ($content['orderedItems'] as $status) {
+				$users = array();
+				$this->items[] = $this->parseItem($status, $users);
+			}
+		} else {
+			throw new \Exception('Unexpected response from server.');
+		}
+	}
+
+	protected function parseItem($content, &$users) {
+		$item = array();
+		switch ($content['type']) {
+			case 'Announce': // boost
+				if ($this->getInput('noboost')) {
+					return null;
+				}
+				// We fetch the boosted content.
+				try {
+					$rtContent = json_decode(getContents($content['object'], self::AP_HEADER), true);
+					if ($rtContent['attributedTo'] && isset($users[$rtContent['attributedTo']])) {
+						$item['author'] = $users[$rtContent['attributedTo']];
+						$item['title'] = 'Shared a status by ' . $item['author'] . ': ';
+						$item = $this->parseObject($rtContent, $item);
+					} else {
+						// We fetch the author, since we cannot always assume the format of the URL.
+						$user = json_decode(getContents($rtContent['attributedTo'], self::AP_HEADER), true);
```

Comment:
I assume this does not typically change more than once a day.
Instead of using the $users array, have you considered using BridgeAbstract::saveCacheValue so that it's preserved across calls? This should help decrease load on the ActivityPub instance.

Reply:
The instance name and the author's username should not change at all.